modules/tensor_mechanics/src/materials/ComputeLagrangianStrain.C

```diff
@@ -0,0 +1,213 @@
+
+  params.addRequiredCoupledVar("displacements", "Displacement variables");
+  params.addParam<bool>(
+      "large_kinematics", false, "Use large displacement kinematics in the kernel.");
+  params.addParam<std::vector<MaterialPropertyName>>("eigenstrain_names",
+                                                     "List of eigenstrains to account for");
+  params.addCoupledVar("macro_gradient", "Optional scalar field with the macro gradient");
+
+  // We rely on this *not* having use_displaced mesh on
+  params.suppressParameter<bool>("use_displaced_mesh");
+
+  return params;
+}
+
+ComputeLagrangianStrain::ComputeLagrangianStrain(const InputParameters & parameters)
+  : DerivativeMaterialInterface<Material>(parameters),
+    _ndisp(coupledComponents("displacements")),
+    _disp(3),
+    _grad_disp(3),
+    _ld(getParam<bool>("large_kinematics")),
+    _eigenstrain_names(getParam<std::vector<MaterialPropertyName>>("eigenstrain_names")),
+    _eigenstrains(_eigenstrain_names.size()),
+    _eigenstrains_old(_eigenstrain_names.size()),
+    _total_strain(declareProperty<RankTwoTensor>("total_strain")),
+    _total_strain_old(getMaterialPropertyOld<RankTwoTensor>("total_strain")),
+    _mechanical_strain(declareProperty<RankTwoTensor>("mechanical_strain")),
+    _mechanical_strain_old(getMaterialPropertyOld<RankTwoTensor>("mechanical_strain")),
+    _strain_increment(declareProperty<RankTwoTensor>("strain_increment")),
+    _def_grad(declareProperty<RankTwoTensor>("deformation_gradient")),
+    _def_grad_old(getMaterialPropertyOld<RankTwoTensor>("deformation_gradient")),
+    _inv_df(declareProperty<RankTwoTensor>("inv_inc_def_grad")),
+    _inv_def_grad(declareProperty<RankTwoTensor>("inv_def_grad")),
+    _detJ(declareProperty<Real>("detJ")),
+    _macro_gradient(isCoupledScalar("macro_gradient", 0) ? coupledScalarValue("macro_gradient")
+                                                         : _zero),
+    _homogenization_contribution(declareProperty<RankTwoTensor>("homogenization_contribution"))
+{
+  // Setup eigenstrains
+  for (unsigned int i = 0; i < _eigenstrain_names.size(); i++)
+  {
+    _eigenstrains[i] = &getMaterialProperty<RankTwoTensor>(_eigenstrain_names[i]);
+    _eigenstrains_old[i] = &getMaterialPropertyOld<RankTwoTensor>(_eigenstrain_names[i]);
+  }
+}
+
+void
+ComputeLagrangianStrain::initialSetup()
+{
+  // Grab the actual displacements
+  for (unsigned int i = 0; i < _ndisp; i++)
+  {
+    _disp[i] = &coupledValue("displacements", i);
+    _grad_disp[i] = &coupledGradient("displacements", i);
+  }
+
+  // All others zero (so this will work naturally for plane strain problems)
+  for (unsigned int i = _ndisp; i < 3; i++)
+  {
+    _disp[i] = &_zero;
+    _grad_disp[i] = &_grad_zero;
+  }
+}
+
+void
+ComputeLagrangianStrain::initQpStatefulProperties()
+{
+  _total_strain[_qp].zero();
+  _mechanical_strain[_qp].zero();
+  _strain_increment[_qp].zero();
+  _def_grad[_qp] = RankTwoTensor::Identity();
```

Comment:
```suggestion
  _def_grad[_qp].setToIdentity();
```
Same suggestion for the following two lines.